Simplify create_readers

pquentin opened the pull request:
This builds upon #1657. The second commit shows a bug in the current implementation of create_readers. It was hidden by the fact we multiply by num_clients which is often more than 1 in multi-corpora implementations.

pquentin commented on esrally/track/params.py:
```diff
@@ -953,39 +953,53 @@ def create_readers(
                     start_client_index,
                     end_client_index,
                     corpus.name,
                 )
```
Good catch, thanks. Fixed in [`a9c0ce7` (#1659)](https://github.com/elastic/rally/pull/1659/commits/a9c0ce76e2d2b5945d9a5e0d79bccbe0ac5874b3) and added a test.

danielmitterdorfer commented on esrally/track/params.py:
```diff
@@ -953,39 +953,54 @@ def create_readers(
+    staggered_readers = []
+    while total_readers > 0:
+        for reader_queue in readers:
+            if reader_queue:
```
I *think* this is only needed because we eagerly add the dequeue to `readers`? If we'd only add it when there are actually documents that need to be read, we could get rid of this if.

pquentin replied:
No, actually this is needed because all queues don't necessarily have the same length. Say I have the following readers list of deques: [[doc1], [doc1, doc2]] and total_reads = 3. After one iteration I will have the following readers: [[], [doc2]] and total_reads = 1. I need to skip the empty deque.

I tried to explain this in [`8f05ed2` (#1659)](https://github.com/elastic/rally/pull/1659/commits/8f05ed285622ebc96940753e07ec418890ca16b0).

danielmitterdorfer commented on esrally/track/params.py:
```diff
@@ -953,39 +953,54 @@ def create_readers(
+            reader = create_reader(
+                docs, offset, num_lines, num_docs, batch_size, bulk_size, id_conflicts, conflict_probability, on_conflict, recency
+            )
+            readers[-1].append(reader)
```
How about we assign the dequeue to a variable? This might be easier to read?

pquentin replied:
Done in [`6870764` (#1659)](https://github.com/elastic/rally/pull/1659/commits/6870764ad9453f5ec8e54265bcc243c6ee4fafa2)

danielmitterdorfer commented on esrally/track/params.py:
```diff
@@ -953,39 +953,54 @@ def create_readers(
-    return readers
+                continue
+
+            target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
```
This and the following two lines add a bit of noise in a quite complicated function. Maybe we can help the reader by moving this into a separate function?

pquentin replied:
So this was only needed for the debug logs, which gave me one more reason to remove them, which I have done in [`a26e18c` (#1659)](https://github.com/elastic/rally/pull/1659/commits/a26e18c84092601c1f116bbbc11b05a81a9585ce)

danielmitterdorfer reviewed (commented):
I only wanted to leave a few high-level comments: For better or worse, bulk-indexing is among the most complex parts of Rally so IMHO it pays off to be a bit more explicit than usual when doing changes. I'm thinking specifically of:

* Adding more comments on the algorithm in this function. What is it doing on a high-level?
* Simplifying the key functionality as much as possible, e.g. do we need the debug log statements?
* Simplifying control structures, e.g. can we reimplement this so we don't need a `continue` statement?
* Can we pick even more descriptive variable names (e.g. what differentiates `readers` from `staggered_readers`? The similar name implies e.g. similarity in data structure which is not the case)?

pquentin commented:
> * Adding more comments on the algorithm in this function. What is it doing on a high-level?

Done in [`8f05ed2` (#1659)](https://github.com/elastic/rally/pull/1659/commits/8f05ed285622ebc96940753e07ec418890ca16b0). This helped me find a bug which I fixed in [`dca5271` (#1659)](https://github.com/elastic/rally/pull/1659/commits/dca5271bfa7f31c51d1d2cf97044273136b0cac9).

> * Reducing noise as much as possible, e.g. do we need the debug log statements?

As mentioned above, this is done in [`a26e18c` (#1659)](https://github.com/elastic/rally/pull/1659/commits/a26e18c84092601c1f116bbbc11b05a81a9585ce).

> * Simplifying control structures, e.g. can we reimplement this so we don't need a `continue` statement?

Done in [`dcee647` (#1659)](https://github.com/elastic/rally/pull/1659/commits/dcee6470cd79f00b500a2ef6bfe599b378bcb919). This was much easier with the debug logs gone!
 
> * Can we pick even more descriptive variable names (e.g. what differentiates `readers` from `staggered_readers`? The similar name implies e.g. similarity in data structure which is not the case)?

Done in [`8f05ed2` (#1659)](https://github.com/elastic/rally/pull/1659/commits/8f05ed285622ebc96940753e07ec418890ca16b0). I also added types in [`6e35a40` (#1659)](https://github.com/elastic/rally/pull/1659/commits/6e35a40fcb85d89083d632cc4610c29c1f91c9f8) to clarify further the contents of the variables.

Can you please take another look?

pquentin commented:
> This LGTM, though I haven't done any local testing, and I'm temporarily without GH access. I can circle back in a few days if you'd like

@DJRickyB I appreciate the offer but you don't have to do this now, you know :)

pquentin commented:
@elasticmachine run rally/it-python38 please

pquentin commented:
Closing/reopening to get a RtD status.

pquentin commented:
@elasticmachine run rally/rally-tracks-compat please now that https://github.com/elastic/rally-tracks/pull/375 was merged.
